linkcheck/checker/urlbase.py

```diff
@@ -383,15 +383,16 @@ def build_url_parts (self):
         Also checks for obfuscated IP addresses.
         """
         # check userinfo@host:port syntax
-        self.userinfo, host = urllib.parse.splituser(self.urlparts[1])
-        port = urlutil.default_ports.get(self.scheme, 0)
-        host, port = urlutil.splitport(host, port=port)
+        o = urllib.parse.urlparse("//%s" % self.urlparts[1])
+        self.userinfo = ":".join([x for x in (o.username, o.password) if x is not None])
```

Comment:
This seems weird to me, so if username is None and password is not, we produce self.userinfo = password.

And if password is None and username is not, we produce self.userinfo = username.

I had to go read the stdlib sources to figure out that if password is not None, then username is guaranteed to be not None as well.

(I also checked if urlparse does any %-decoding, and no, it doesn't.)

Reply:
i agree this could be more readable. i can barely parse it now, and so i will definitely have trouble parsing it later. could we maybe expan that loop or at least explain what we're doing here or why?

what was wrong with splituser anyways?

in any case, this was more readable before and now seems a little weird, so i would like to keep the readability, if possible.

Reply:
> what was wrong with splituser anyways?

it emits deprecation warnings now:

https://github.com/python/cpython/blob/3.8/Lib/urllib/parse.py#L1038-L1042

Reply:
hum. well it's still not pretty. :)

Reply:
yeah, it feels as if we're doing something wrong.

self.urlparts is already the result of urlparse, isn't it?  in which case maybe we have a self.urlparts.username and self.urlparts.password already?

Reply:
> self.urlparts is already the result of urlparse, isn't it? in which case maybe we have a self.urlparts.username and self.urlparts.password already?

No, but only because we use strformat.url_unicode_split(), which converts the ParseResult object into a plain list.

url_unicode_split is not necessary in a Python 3 only world and we should remove it.

I think I'm going to retract my approval for this PR.  Better things are possible!

Reply:
Yes. The review comments are making me think it is pretty fragile and the effort would be better put towards reviewing the entire file. Something post 10? The deprecations only came in with Python 3.8 so hopefully will be still supported for a while.

Reply:
Yep, there's no rush!